### DISCUSSION TITLE
support for JBIG2 compression

### DISCUSSION THREAD

**Discussion by amigalemming** [ImageMagick/ImageMagick#6701] [Contribute]
So far I scan black/white documents with high resolution and convert them to PDF using imagemagick:
~~~~
convert page*.png -compress fax document.pdf
~~~~
This already gives me good compression results. I would like to also get JBIG2 compression by replacing `fax` by `jbig2`. (When I currently do this the result file size looks like uncompressed output.)
So far I have to use [jbig2enc](https://github.com/agl/jbig2enc) which is cumbersome to install and to use.

**Comment by mikayla-grace** [comment]
ImageMagick does not support JBIG2 compression for PDF image files.

**Comment by amigalemming** [comment]
On Thu, 27 Jul 2017, mikayla-grace wrote:

> ImageMagick does not support JBIG2 compression for PDF image files.

That's why I propose it as a new feature.

**Comment by bdheeman** [comment]
I also had been using IM to grayscale, clean, deskew, trim and center scanned pages into black/white 1-bit PBM files and adding the same to PDF file, finally. I attempted to try and experiment with various compression and encoding techniques until I stumbled upon [jbig2enc](https://github.com/agl/jbig2enc), which did the job nicely and is lot less resource hungry.

I'm of an opinion, if someone is already doing something well, why re-invent the wheel? But, that's an opinion only and the opinions can be wrong ;-)

See also: #841

**Comment by Artoria2e5** [comment]
I think JBIG2 is there now.

On the official 7.0.8-42 (Windows Q16-x64) build:

```
> magick -list compress
B44A
B44
BZip
DXT1
DXT3
DXT5
Fax
Group4
JBIG1
JBIG2
JPEG2000
JPEG
LosslessJPEG
Lossless
LZMA
LZW
None
Piz
Pxr24
RLE
RunlengthEncoded
WebP
ZipS
Zip
Zstd
```

**Comment by amigalemming** [comment]
On Fri, 26 Apr 2019, Mingye Wang wrote:

> I think JBIG2 is there now.

But it seems to be a lie.

I have a JBIG2 compressed PDF, generated with jbig2enc:

$ ll test.pdf
-rw------- 1 thielema users 54990 Mär 30 07:51 test.pdf

$ pdfimages -png test.pdf page
$ convert page-*.png -compress fax test-fax.pdf
$ convert page-*.png -compress jbig2 test-jbig2.pdf
$ ll test-*.pdf
-rw------- 1 lemming users  164283 Apr 26 18:45 test-fax.pdf
-rw------- 1 lemming users 1343381 Apr 26 18:44 test-jbig2.pdf

"JBIG2" compression generated by 'convert' is 20 times as large as the 
original JBIG2 compressed file, and still 10 times as large as the FAX 
compressed PDF. Even if it is JBIG2 compressed, it is badly compressed. 
That's why I opened the ticket.

**Comment by josch** [comment]
Try opening your `test-jbig2.pdf` with a text editor and look for a line starting with `/Filter`. Instead of `/JBIG2Decode` the filter will be `/RunLengthDecode` -- even though you told imagemagick to use jbig2. This is why your files are so big. Indeed, even though instructed otherwise, imagemagick is not creating a pdf with the image encoded with jbig2 compression. And this is a bug.

**Comment by urban-warrior** [comment]
Not a bug, ImageMagick does not support JBIG2 compression for PDF image files.

**Comment by josch** [comment]
Okay, sorry: feature request. :)

**Comment by olcc** [comment]
Why is `convert -list compress` mentioning `JBIG2` if it actually cannot be used for PDF files? I wish the JBIG2 compression could be used for creating PDF files with ImageMagick. Currently, only `Fax` or `Group4` work well for integrating my monochrome images into PDF files.

**Comment by fmw42** [comment]
Each format has specific compressions that can be used for that format. Apparently PDF does not use JBIG2. But I am not a PDF expert to know if that is possible or not.

**Comment by josch** [comment]
> Each format has specific compressions that can be used for that format. Apparently PDF does not use JBIG2. But I am not a PDF expert to know if that is possible or not.

The fact that the PDF format can store images as JBIG2 became infamously known via a bug in xerox scanners which used an incorrect JBIG2 encoding of scanned documents when saving them as PDF such that single digits of the document got swapped out: https://www.dkriesel.com/en/blog/2013/0802_xerox-workcentres_are_switching_written_numbers_when_scanning So yes, PDF 100% is able to store images in JBIG2 format.

**Comment by fmw42** [comment]
My answer was more about why a format would be listed and could not be use for a particular application. Formats use different compressions.  JBIG2 compression is useful for other formats or for other applications. So it is listed. There is not current way that I know to as for a list of compression for a give image format.

As I said, I am not a PDF expert, so do not know the reason IM does not use it for PDFs.

**Comment by olcc** [comment]
So, it is a pity, the situation is the following:
- ImageMagick can create JBIG images...
- but cannot integrate them into a PDF file

**Comment by urban-warrior** [comment]
Can you post links to a few sample PDF's that include JBIG-compressed images?  With them, we can work toward supporting JBIG compression in PDF's within ImageMagick.

**Comment by josch** [comment]
The link I posted earlier has PDF files as they come out of the Xerox WorkCentre 7535. Those use `/Filter /JBIG2Decode`. Here is the direct link to a PDF from that article:

https://www.dkriesel.com/_media/blog/2013/7535-7pt.pdf

**Comment by olcc** [comment]
Google Books use JBIG2 massively for producing PDF files with a small size. Here is a [page](https://github.com/ImageMagick/ImageMagick/files/8800702/jbig2_page.pdf) extracted from one of them. It includes JBIG2 images.

**Comment by urban-warrior** [comment]
We can create JBIG images.  What we need is several samples of JBIG-compressed PDF image files so we can reverse-engineer the PDF JBIG requirements.  We have two examples now.  A couple more would be useful.

**Comment by amigalemming** [comment]
Here is an example PDF generated by jbig2enc:
https://github.com/amigalemming/ImageMagick/blob/jbig2-pdf/images/github.pdf

**Comment by raffaelemancuso** [comment]
Any news on this?

**Comment by urban-warrior** [comment]
We added a label "up for grabs" which means we ask the user community to contribute a pull request for this feature.   So far no one has responded.

**Comment by pts** [comment]
FYI The jbig2 command-line tool can create a PDF `/JBIG2Decode` stream. A stripped-down version is available at https://github.com/pts/pdfsizeopt-jbig2/blob/master/c.sh
